### PR TITLE
DCS-1074 moved to new path and availability service for amend journey

### PR DESCRIPTION
**Pull request by louisenorris**
Screenshot 2021-07-14 at 13 38 40

Amend journey adjusted to follow updated path as shown in above diagram. Video Link available and Select rooms page removed and substituted with an updated Change video link booking page and Check and confirm your booking page. URLs also updated in accordance with updates to the BVL URL spreadsheet. 

Video link not available page to be updated with alternative video link booking options in a future ticket.

### CODE REVIEW

File: backend/routes/amendBooking/changeVideoLinkController.test.ts [hunk 1]
```diff
@@ -1,18 +1,18 @@
 import moment from 'moment'
-import ChangeVideoLinkBookingController from './changeVideoLinkBookingController'
-import { BookingService, LocationService, AvailabilityCheckServiceV1 } from '../../services'
+import ChangeVideoLinkController from './changeVideoLinkController'
+import { BookingService, LocationService, AvailabilityCheckServiceV2 } from '../../services'
 
-import { BookingDetails, RoomAvailability } from '../../services/model'
+import { BookingDetails, RoomAvailabilityV2 } from '../../services/model'
```
Comment (andrewrlee) [inline backend/routes/amendBooking/changeVideoLinkController.test.ts:5]: would be nice to drop the V2 suffix in a future PR

File: backend/routes/amendBooking/confirmUpdatedBookingController.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  ) {}
+
+  public view(): RequestHandler {
+    return async (req, res) => {
+      const { bookingId } = req.params
+      const errors = req.flash('errors') || []
+      const [input] = req.flash('input')
+
+      const update = getUpdate(req)
+      if (!update) {
```
Comment (andrewrlee) [inline backend/routes/amendBooking/confirmUpdatedBookingController.ts:21]: Maybe could move this to middleware and apply across a variety of routes like we do with the `ensureNewBookingPresentMiddleware` (possible follow up?)

https://github.com/ministryofjustice/hmpps-book-video-link/blob/main/backend/routes/createBooking/index.ts#L19

File: backend/routes/amendBooking/confirmUpdatedBookingController.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            mainCourtHearingTime: `${update.startTime.format(MOMENT_TIME)} to ${update.endTime.format(MOMENT_TIME)}`,
+            prisonRoomForCourtHearing: roomFinder.prisonRoom(update.mainLocation),
+            'pre-court hearing briefing': update.preRequired && formatTimes(preAppointmentTimes(update.startTime)),
+            'prison room for pre-court hearing briefing': roomFinder.prisonRoom(update.preLocation),
+            'post-court hearing briefing': update.postRequired && formatTimes(postAppointmentTimes(update.endTime)),
+            'prison room for post-court hearing briefing': roomFinder.prisonRoom(update.postLocation),
+          },
+        },
+        form,
+        changeBookingLink: `/${bookingDetails.agencyId}/offenders/${bookingDetails.offenderNo}/add-court-appointment`,
```
Comment (andrewrlee) [inline backend/routes/amendBooking/confirmUpdatedBookingController.ts:54]: Think this url links to the create flow?

I think this should be the change booking url: `/change-video-link/${bookingId}`

File: backend/routes/amendBooking/confirmUpdatedBookingController.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        return res.redirect(`/booking-details/${bookingId}`)
+      }
+
+      const bookingDetails = await this.bookingService.get(res.locals, parseInt(bookingId, 10))
+
+      const courtUpdate = await this.locationService.getVideoLinkEnabledCourt(res.locals, update.courtId)
+
+      const roomFinder = await this.locationService.createRoomFinder(res.locals, bookingDetails.agencyId)
+
+      const form = input ? Comment(input) : { comment: bookingDetails.comments }
```
Comment (andrewrlee) [inline backend/routes/amendBooking/confirmUpdatedBookingController.ts:31]: Realise its not consistent with the other forms but could consider getting rid of the `Comment` object as only used to wrap and splat and doesn't really provide a lot of benefit.

could even: 
```
const comment = input ? input.comment : bookingDetails.comments
```

and then in res.render add `comment` directly

File: backend/routes/amendBooking/confirmUpdatedBookingController.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    return async (req, res) => {
+      const { bookingId } = req.params
+      if (req.errors) {
+        req.flash('errors', req.errors)
+        req.flash('input', req.body)
+        return res.redirect(`/confirm-updated-booking/${bookingId}`)
+      }
+
+      const update = getUpdate(req)
+      const comment = Comment(req.body)
```
Comment (andrewrlee) [inline backend/routes/amendBooking/confirmUpdatedBookingController.ts:69]: same here?

```
const { comment } = req.body
...
{ ...update, comment }
```

Think this `Comment` form abstraction would be worth it's weight more if it was doing some kind of type conversion but as it's a string probably not worth it... though might change if we look at using something like class validator.

File: backend/routes/amendBooking/index.ts [hunk 6]
```diff
@@ -3,53 +3,48 @@ import { Services } from '../../services'
...
+  const roomNoLongerAvailable = new RoomNoLongerAvailableController()
+  const confirmUpdatedBooking = new ConfirmUpdatedBookingController(bookingService, locationService)
   const confirmation = new ConfirmationController(bookingService)
   const changeComments = new ChangeCommentsController(bookingService)
 
   const router = express.Router({ mergeParams: true })
 
-  router.get('/start-change-date/:bookingId', asyncMiddleware(changeVideoLinkBooking.start()))
-  router.get('/change-video-link-date-and-time/:bookingId', asyncMiddleware(changeVideoLinkBooking.view()))
+  router.get('/start-change-date/:bookingId', asyncMiddleware(changeVideoLink.start()))
```
Comment (andrewrlee) [inline backend/routes/amendBooking/index.ts:30]: might be good to change this url and it's associated reference in the search page, think this should be `/start-change-booking/:bookingId` ?

File: backend/services/availabilityCheckServiceV2.ts [hunk 7]
```diff
@@ -48,7 +40,7 @@ export default class AvailabilityCheckServiceV2 implements AvailabilityStatusChe
 
   public async getAvailabilityStatus(
     context: Context,
-    request: AvailabilityRequest,
+    request: AvailabilityRequestV2,
     selectedRooms: SelectedRooms
   ): Promise<AvailabilityStatus> {
     const { isAvailable } = await this.getAvailability(context, {
```
Comment (andrewrlee) [inline backend/services/availabilityCheckServiceV2.ts:46]: presumably we can get rid of `SelectedRooms` argument now, as the rooms are present in the `AvailabilityRequestV2`

maybe something like?:
`const { isAvailable } = await this.getAvailability(context, request)`

File: views/amendBooking/confirmUpdatedBooking.njk [hunk 8]
```diff
@@ -34,38 +21,30 @@
...
-            {% if preRequired %}
-                {{ appointment({
-                    description: "Prison room for pre-court hearing briefing", 
-                    name: "preLocation",
-                    location: form.preLocation,
-                    locations: preLocations
-                    }, 
-                    errors) }}
-            {% endif %}
+        <div class="govuk-grid-row">
```
Comment (andrewrlee) [inline views/amendBooking/confirmUpdatedBooking.njk:35]: meh.. don't think you need  to wrap each summary list in its own two divs:

```
<div><div>govukSummaryList()<div><div>
<div><div>govukSummaryList()<div><div>
```
 could be
```
<div>
  <div>
     govukSummaryList() 
     govukSummaryList()
  <div>
<div>
```

(is how create booking works)

File: backend/routes/amendBooking/confirmUpdatedBookingController.ts (hunk 3 above)
Comment (louisenorris) [reply backend/routes/amendBooking/confirmUpdatedBookingController.ts:54]: Ah thanks - updated

File: backend/routes/amendBooking/changeVideoLinkController.test.ts (hunk 1 above)
Comment (louisenorris) [reply backend/routes/amendBooking/changeVideoLinkController.test.ts:5]: So all good to drop the V2 suffix on AvailabilityServiceV2 in the next commit on this PR?

File: backend/routes/amendBooking/confirmUpdatedBookingController.ts (hunk 4 above)
Comment (louisenorris) [reply backend/routes/amendBooking/confirmUpdatedBookingController.ts:31]: Cool, updated as suggested

File: backend/routes/amendBooking/confirmUpdatedBookingController.ts (hunk 5 above)
Comment (louisenorris) [reply backend/routes/amendBooking/confirmUpdatedBookingController.ts:69]: Cool, updated as suggested

File: backend/routes/amendBooking/index.ts (hunk 6 above)
Comment (louisenorris) [reply backend/routes/amendBooking/index.ts:30]: Updated as suggested

File: views/amendBooking/confirmUpdatedBooking.njk (hunk 8 above)
Comment (louisenorris) [reply views/amendBooking/confirmUpdatedBooking.njk:35]: Updated as suggested

File: backend/services/availabilityCheckServiceV2.ts (hunk 7 above)
Comment (louisenorris) [reply backend/services/availabilityCheckServiceV2.ts:46]: Cool, updated as suggested

File: backend/routes/amendBooking/changeVideoLinkController.test.ts (hunk 1 above)
Comment (andrewrlee) [reply backend/routes/amendBooking/changeVideoLinkController.test.ts:5]: Yep, next PR is fine

**Review by andrewrlee — APPROVED** [review]
Nice! 

LGTM

**Pull request merged**